### ISSUE TITLE
apply_redactions causes part of the page content to be hidden / transparent

### ISSUE THREAD

**Issue by beeing** [pymupdf/PyMuPDF#3751]
### Description of the bug

I'm adding a redaction region to a part of the PDF, but after calling apply_redactions(), one side of the entire page goes missing (opened in macOS Preview app or Safari).

Further inspection reveals that the text is not missing, as it is selectable and can be copied out properly. It is either the text has been masked / hidden, but I could not find out how to check further (sorry, my limited knowledge on PDF structure).

![Untitled](https://github.com/user-attachments/assets/7d214d49-5a65-400a-9309-7c1027421fab)

The media, crop, art, bleed, trim boxes all looks fine before and after the redactions. In fact, I'm trying to check if there's other paths, objects that may be causing it but there's nothing.

Note that I'm not able to share the actual PDF but it was generated from Puppeteer / Chromium (PDF ver 1.7).

Thanks in advance for looking into this.

### How to reproduce the bug

1. Generate PDF from Chromium / Puppeteer
2. Add redaction of any size eg. `(0,0,1,1)` and call `apply_redactions()`
3. Open the PDF in Preview App.

### PyMuPDF version

1.24.9

### Operating system

MacOS

### Python version

3.9

**Comment by Kyrylo-Hrytsenko** [comment]
I faced the same problem.
_**Note**_: this problem does not exist on version 1.23.9, all higher versions have it

**Comment by JorjMcKie** [comment]
As always: please provide an example PDF! There is no way to otherwise deal with this post.

**Comment by beeing** [comment]
I've just tested and it works on older version (up to pymupdf-1.23.26).

Perhaps easier to compare the commits since https://github.com/pymupdf/PyMuPDF/commit/a868c0a556e39198549e4e139534bb12b2623c5d until HEAD.

**Comment by Kyrylo-Hrytsenko** [comment]
@JorjMcKie 
Here is an example
[original.pdf](https://github.com/user-attachments/files/16511592/original.pdf)
[redacted.pdf](https://github.com/user-attachments/files/16511591/redacted.pdf)

Plese, open redacted file with Preview app, it will looks like this
<img width="1090" alt="Screenshot 2024-08-06 at 3 38 40 PM" src="https://github.com/user-attachments/assets/fbece197-5d4a-47a5-81ee-10cc47461851">

The code for redaction looks like this:

```
pdfIn = fitz.open(input_file)

out_buffer = BytesIO()

page = pdfIn[0]

page.add_redact_annot([0,0,100,100], text=None, fill=(0, 0, 0))
page.add_redact_annot([100,100,200,200], text=None, fill=(0, 0, 0))
page.add_redact_annot([200,200,300,300], text=None, fill=(0, 0, 0))
page.add_redact_annot([300,300,400,400], text=None, fill=(0, 0, 0))

page.apply_redactions()

pdfIn.save(out_buffer, garbage=3, deflate=True)
pdfIn.close()

with open(output_file, mode='wb') as f:
    f.write(out_buffer.getbuffer())
f.close()

```

**Comment by JorjMcKie** [comment]
@Kyrylo-Hrytsenko Thanks, I did.

I executed the script and found no problem at all using v1.24.9. I modified the script somewhat so redaction rectangles are visible and erased areas are not filled:
```python
import pymupdf

print(pymupdf.version)
pdfIn = pymupdf.open("original.pdf")

page = pdfIn[0]
rects = (
    [0, 0, 100, 100],
    [100, 100, 200, 200],
    [200, 200, 300, 300],
    [300, 300, 400, 400],
)
for r in rects:
    page.draw_rect(r, color=(1, 0, 0))
    page.add_redact_annot(r)

page.apply_redactions()

pdfIn.ez_save("output.pdf")
```
Gives this correct result: 
[output.pdf](https://github.com/user-attachments/files/16636124/output.pdf)

**Comment by Kyrylo-Hrytsenko** [comment]
@JorjMcKie 
Your result file looks like this for me in the Preview app: 
<img width="1094" alt="Screenshot 2024-08-16 at 1 59 20 PM" src="https://github.com/user-attachments/assets/3a58ee53-e81e-4c20-9744-6b70b70e26a4">

Notes:
 - I highlighted some text with the mouse to show that it is still present in the PDF, but for some reason, it's hidden.
 - There is no issue when I open this file in Chrome or any other app.
 - I checked version 1.23.9, and there was no such issue. It seems to have started happening in version 1.24.0.

Does the output.pdf look normal when you open it in the 'Preview' application?

**Comment by JorjMcKie** [comment]
I do not use or have Preview.
My file is displayed in all PDF viewers like Adobe Acrobat, Foxit, Nitro, PDF XChange, evince (Linux). So all authoritative applications behave correctly.
No idea what is wrong with Preview.

**Comment by jamie-lemon** [comment]
Can confirm also see this problem in Preview. However it is fine when I open in Adobe Acrobat. To me this feels like a Preview  rendering bug. I would submit a bug to Apple if that is possible!

**Comment by JorjMcKie** [comment]
@jamie-lemon absolutely correct! I was about to write a similar comment.
We will now close this issue.

**Comment by Kyrylo-Hrytsenko** [comment]
@jamie-lemon @JorjMcKie 
I don't think it's Preview bug, here is why:
  - For me, it happens not only with Preview but also with the react-pdf-viewer library at least
  - With an older version of your library everything works fine, which means something was changed and caused this issue
  - Original files (before redaction) render correctly with Preview and with react-pdf-viewer, which means something in the redaction process causes this issue.

**Comment by yuhuang-cst** [comment]
> @jamie-lemon @JorjMcKie I don't think it's Preview bug only, here is why:
> 
> * For me, it happens not only with Preview but also with the react-pdf-viewer library at least
> * With an older version of your library everything works fine, which means something was changed and caused this issue
> * Original files (before redaction) render correctly with Preview and with react-pdf-viewer, which means something in the redaction process causes this issue.

@JorjMcKie @jamie-lemon 
In addition to **Mac Preview**, **Safari**, **UPDF**, and **PDF Expert** also fail to display `output.pdf` correctly.

**Comment by jamie-lemon** [comment]
This is a strange bug - I thought it might be related to the content on page 1, but if I simplify things, target the 2nd page with an area redaction with:
```
import pymupdf

print(pymupdf.version)
pdfIn = pymupdf.open("orginal.pdf")

page = pdfIn[1] #2nd page
rects = (
    [0, 0, 100, 100],

)
for r in rects:
    page.draw_rect(r, color=(1, 0, 0))
    page.add_redact_annot(r)

page.apply_redactions()

pdfIn.ez_save("redacted.pdf")
```
Then I get:
<img width="645" alt="Screenshot 2024-08-16 at 18 05 59" src="https://github.com/user-attachments/assets/120c11a9-8e84-408b-9f06-65773ed8b189">

I also noticed that it doesn't matter how big the area redaction, I could do this:

```
rects = (
    [0, 0, 0, 0],
)

And achieve the same resulting problem with the left hand side of the page.

Testing with other documents, redacting and viewing in Preview I don't find this issue at all, so I think there must be something *very* specific to this document which will need further research.

**Comment by Kyrylo-Hrytsenko** [comment]
@jamie-lemon 

>  so I think there must be something very specific to this document which will need further research.

Totally agree. Only a small number of my documents have this bug. I didn't even plan to write to you but then noticed that this is happening not only to me and the bug was already created, so I added my example as well.

**Comment by jamie-lemon** [comment]
@Kyrylo-Hrytsenko Much appreciated!

**Comment by jamie-lemon** [comment]
This is the simplest case I could find - I made this PDF in Adobe Acrobat, then took it into Preview and then did "Export" as a new PDF. 

[preview-made.pdf](https://github.com/user-attachments/files/16641540/preview-made.pdf)

When you redact with PyMuPDF the logo disappears when you view it in Preview, e.g.

<img width="704" alt="Screenshot 2024-08-16 at 21 24 35" src="https://github.com/user-attachments/assets/18e3c048-026c-461d-9dc4-b4f34df070f3">

**Comment by jamie-lemon** [comment]
So it seems if the PDF is made in Preview then this might have something to do with the problem.

**Comment by yuhuang-cst** [comment]
The issue I am encountering is that if apply_redactions is used, the vector graphics on the page all move to the bottom left corner in **Preview**, **Safari**, **UPDF**, and **PDF Expert**, whereas they display correctly in **Chrome**, **Adobe Acrobat Reader**, and **WPS**. Here is the code:

```python
import fitz
doc = fitz.open('origin.pdf')
page = doc.load_page(0)
page.add_redact_annot((0, 0, 0 ,0), fill=False)
page.apply_redactions()
doc.ez_save('apply_redaction.pdf')
doc.close()
```

[origin.pdf](https://github.com/user-attachments/files/16643236/origin.pdf)
[apply_redaction.pdf](https://github.com/user-attachments/files/16643237/apply_redaction.pdf)
<img width="718" alt="image" src="https://github.com/user-attachments/assets/93128011-a80b-4913-b7ef-06553c9fcd72">

The `origin.pdf` is from the second page of the AlphaGo paper: https://www.nature.com/articles/nature16961

**Comment by yuhuang-cst** [comment]
The PDF generated with PyMuPDF version 1.23.26 displays the vector graphics correctly in Preview (although the image in the top right corner is also missing). However, starting from version 1.24.0, there is a bug where the vector graphics are moved to the bottom left corner.
[apply_redaction_1.23.26.pdf](https://github.com/user-attachments/files/16643282/apply_redaction_1.23.26.pdf)
[apply_redaction_1.24.0.pdf](https://github.com/user-attachments/files/16643284/apply_redaction_1.24.0.pdf)

**Comment by yuhuang-cst** [comment]
> It seems that primarily Mac-based tools have problems with redacted PDFs that have been created with Preview. I am experimenting with the MuPDF development version 1.25.0. The current PyMuPDF v1.24.9 uses MuPDF v1.24.8.
> 
> When creating and applying annotations using PyMuPDF 1.24.9 with MuPDF 1.25.0 I do no longer see the error using the Firefox browser - which does behave awkwardly as all those Mac apps.
> 
> I am attaching the produced output.pdf inviting Mac users to access it with their Preview on Mac: [output.pdf](https://github.com/user-attachments/files/16644815/output.pdf)

<img width="1277" alt="image" src="https://github.com/user-attachments/assets/4e378e70-5eb3-4650-b64c-d3e8fec6f194">
It seems that this bug still exists in Mac Preview.

**Comment by JorjMcKie** [comment]
@yuhuang-cst thanks for the feedback anyway

**Comment by jamie-lemon** [comment]
Can also confirm that the bug doesn't exist in PyMuPDF version 1.23.9

**Comment by JorjMcKie** [comment]
I have submitted a problem report in MuPDF's system here:https://bugs.ghostscript.com/show_bug.cgi?id=707966

**Comment by K8S666** [comment]
Report the same issue。

**Comment by julian-smith-artifex-com** [comment]
Apparent MuPDF master branch has a fix, so PyMuPDF itself will be fixed when we make a release with MuPDF-1.25.0.

However i don't currently know when MuPDF_1.25.0 will be released.

**Comment by jamie-lemon** [comment]
@julian-smith-artifex-com @JorjMcKie Let's release soon to hopefully fix this one! Related I think: https://github.com/pymupdf/PyMuPDF/issues/3751

**Comment by julian-smith-artifex-com** [comment]
Fixed in PyMuPDF-1.25.0.
